lib/remoteproc/remoteproc_virtio.c

```diff
@@ -114,22 +129,20 @@ static void rproc_virtio_set_features(struct virtio_device *vdev,
 	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
 	vdev_rsc = rpvdev->vdev_rsc;
 	io = rpvdev->vdev_rsc_io;
-	/* TODO: shall we set features based on the role ? */
 	metal_io_write32(io,
-			 metal_io_virt_to_offset(io, &vdev_rsc->dfeatures),
+			 metal_io_virt_to_offset(io, &vdev_rsc->gfeatures),
 			 features);
 	rpvdev->notify(rpvdev->priv, vdev->index);
 }
-#endif
 
 static uint32_t rproc_virtio_negotiate_features(struct virtio_device *vdev,
 						uint32_t features)
 {
-	(void)vdev;
-	(void)features;
-
+	/* Assume the master support all slave features */
+	rproc_virtio_set_features(vdev, features);
```

Comment:
How about do gfeatures = (features & with vdev_rsc->dfeatures) in this function for master?
I think the comments can be removed, from this function, instead, the comments can be put to the rproc_create_vdev().

We have a gap in feature negotiation. I am thinking:
* Opt 1:
to have feature negotiate called for slave too.
The feature negotiate can only be called for slave when the DRIVER_OK is set.
The final features will be dfeatures & gfeatures
The negotiate_features function will set the local features variable. The future get_features will not need to check the virtio config space, but just return the local features variable value.

* opt 2:
get_features() functions always get dfeatures and gfeatures from virtio config space and return (dfeatures & gfeatures).

Maybe Opt2 is better

Reply:
> How about do gfeatures = (features & with vdev_rsc->dfeatures) in this function for master?

Good point.

> I think the comments can be removed, from this function, instead, the comments can be put to the rproc_create_vdev().
> 

Yes, I should put there.

> We have a gap in feature negotiation. I am thinking:
> 
> * Opt 1:
>   to have feature negotiate called for slave too.
>   The feature negotiate can only be called for slave when the DRIVER_OK is set.
>   The final features will be dfeatures & gfeatures
>   The negotiate_features function will set the local features variable. The future get_features will not need to check the virtio config space, but just return the local features variable value.
> * opt 2:
>   get_features() functions always get dfeatures and gfeatures from virtio config space and return (dfeatures & gfeatures).
> 
> Maybe Opt2 is better

I think too.